zkvm/src/signature/musig.rs

```diff
@@ -0,0 +1,175 @@
+#![allow(non_snake_case)]
+
+use crate::errors::VMError;
+use crate::signature::VerificationKey;
+use crate::transcript::TranscriptProtocol;
+use curve25519_dalek::constants::RISTRETTO_BASEPOINT_POINT;
+use curve25519_dalek::ristretto::RistrettoPoint;
+use curve25519_dalek::scalar::Scalar;
+use merlin::Transcript;
+
+#[derive(Clone)]
+pub struct PubKey(pub RistrettoPoint);
+
+#[derive(Clone)]
+pub struct PrivKey(pub Scalar);
+
+#[derive(Debug, Clone)]
+pub struct Signature {
+    pub s: Scalar,
+    pub R: RistrettoPoint,
+}
+
+#[derive(Clone)]
+pub struct Message(pub Vec<u8>);
+
+impl Signature {
+    pub fn verify(
+        &self,
+        // The message `m` should already have been fed into the transcript
+        transcript: &Transcript,
+        X_agg: VerificationKey,
+    ) -> Result<(), VMError> {
+        let G = RISTRETTO_BASEPOINT_POINT;
+        let mut transcript = transcript.clone();
+
+        // Make c = H(X_agg, R, m)
+        // The message `m` should already have been fed into the transcript
+        let c = {
+            transcript.commit_point(b"X_agg", &X_agg.0);
+            transcript.commit_point(b"R", &self.R.compress());
+            transcript.challenge_scalar(b"c")
+        };
+
+        let X_agg = match X_agg.0.decompress() {
```

Comment:
i wonder if it makes sense to revamp `VerificationKey` to be decompressed to make usage non-failable. We can keep the compressed copy inside too, so we don't need to recompress when hashing into transcript?.. we can get to this in a separate PR, tho.

Reply:
If `VerificationKey` is being passed around between parties, then it makes sense to be compressed... but for within-ZkVM use, it would be easier for it to be decompressed. I don't have a good sense for the communication requirements for `VerificationKey`s. Maybe a separate PR?